lib/extras/enc/encode.cc

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#include "lib/extras/enc/encode.h"
+
+#include <locale>
+
+#if JPEGXL_ENABLE_APNG
+#include "lib/extras/enc/apng.h"
+#endif
+#if JPEGXL_ENABLE_EXR
+#include "lib/extras/enc/exr.h"
+#endif
+#if JPEGXL_ENABLE_JPEG
+#include "lib/extras/enc/jpg.h"
+#endif
+#include "lib/extras/enc/pgx.h"
+#include "lib/extras/enc/pnm.h"
+#include "lib/jxl/base/printf_macros.h"
+
+namespace jxl {
+namespace extras {
+
+Status SelectFormat(const std::vector<JxlPixelFormat>& accepted_formats,
+                    const JxlBasicInfo& basic_info, JxlPixelFormat* format) {
+  const size_t original_bit_depth = basic_info.bits_per_sample;
+  const auto num_channels =
+      basic_info.num_color_channels + (basic_info.alpha_bits != 0 ? 1 : 0);
+  size_t current_bit_depth = 0;
+  for (const JxlPixelFormat& candidate : accepted_formats) {
+    if (candidate.num_channels != num_channels) continue;
+    const size_t candidate_bit_depth =
+        PackedImage::BitsPerChannel(candidate.data_type);
+    if (
+        // Candidate bit depth is less than what we have and still enough
+        (original_bit_depth <= candidate_bit_depth &&
+         candidate_bit_depth < current_bit_depth) ||
+        // Or larger than the too-small bit depth we currently have
+        (current_bit_depth < candidate_bit_depth &&
+         current_bit_depth < original_bit_depth)) {
+      *format = candidate;
+      current_bit_depth = candidate_bit_depth;
+    }
+  }
+  if (current_bit_depth == 0) {
+    return JXL_FAILURE("no appropriate format found");
+  }
+  if (current_bit_depth < original_bit_depth) {
+    JXL_WARNING("encoding %" PRIuS "-bit original to %" PRIuS " bits",
+                original_bit_depth, current_bit_depth);
+  }
+  return true;
+}
+
+std::unique_ptr<Encoder> Encoder::FromExtension(std::string extension) {
+  std::transform(
+      extension.begin(), extension.end(), extension.begin(),
+      [](char c) { return std::tolower(c, std::locale::classic()); });
```

Comment:
What will happen here if the extension is not valid ASCII?

Reply:
(TBC, "doing something weird" is fine with me, as long as it doesn't die)

Reply:
I don’t think it should die: https://en.cppreference.com/w/cpp/locale/tolower

> Returns the lowercase form of `ch` if one is listed in the locale, otherwise return `ch` unchanged.